### ISSUE TITLE
Stack overflow with very large functions

### ISSUE THREAD

**Issue by stedolan** [ocaml/ocaml#9509]
There's a theoretical issue with stack overflow detection with very large functions, that came up today in a conversation with @mshinwell.

Some OSes (notably, Windows, and certain configurations of threads on Linux/glibc) keep only a single guard page at the end of the stack, and memory just afterwards could be used for other purposes.

If a very large function with a frame size of more than 4k is entered when the stack is about to overflow, it's possible for the first write the function does to skip over the guard page and corrupt something else. Very soon, the function will hit the guard page and a `Stack_overflow` will be detected, but some corruption could occur beforehand.

The fix is fairly straightforward, but must be done per-platform. During the function prologue, when the stack is being expanded, functions with large stacks should do one write to the stack every 4k. This guarantees that they write to the guard page (triggering a fault) before overwriting any information beyond the end of the stack.

**Comment by xavierleroy** [comment]
GCC has the `-fstack-check` option to generate this kind of page-by-page "touch" of stack space.  Note that it is off by default, even though large stack frames are much more common in C than in OCaml (owing to stack-allocated arrays).  So, yes, we could do that in OCaml, but I think we have much more urgent things to do.

**Comment by stedolan** [comment]
Agreed, this is definitely not urgent at all.

**Comment by DemiMarie** [comment]
@xavierleroy it is worth noting that some distributions do turn it on by default.  Stack clashing has been a source of vulnerabilities in setuid binaries, IIRC.

**Comment by DemiMarie** [comment]
Is this a problem for setuid binaries written in OCaml?

**Comment by xavierleroy** [comment]
> Is this a problem for setuid binaries written in OCaml?

You may be thinking of the attack whereas a very large stack frame is created, so large (gigabytes or more) that it jumps over the guard pages for the stack and overlaps with data pages.  This can be achieved relatively easily in C using very large stack-allocated arrays, or even more easily in C 99 with variable-length arrays.

In OCaml, the stack frame only contains scalar values (pointers or numbers), so if you try to build a multi-gigabyte stack frame, the compiler will probably die while compiling your code.

In summary: I'm not too worried about stack clashing that would overwrite data pages.  The worse that could happen here (I think) is a stack overflow that cannot be reported as a Stack_overflow exception and ends up as a "segmentation violation" signal.

**Comment by xavierleroy** [comment]
OCaml 5 has explicit detection of stack overfow, avoiding this issue.  I don't think it's worth an emergency fix in 4.14 LTS.  Closing.
